### ISSUE TITLE
Small inconsistency in the TxSubmission2 miniprotocol spec

### ISSUE THREAD

**Issue by Quantumplation** [input-output-hk/ouroboros-network#4346]
Hi @coot, I've been doing some fun things over on https://github.com/txpipe/pallas, implementing the server-side protocols from Rust so that I can build an archival specialist-node, https://github.com/Quantumplation/cardano-slurp

I've really enjoyed the quality of the networking spec, so thank you and your team for that!  However, I did notice some small inconsistencies that @JaredCorduan suggested I raise here.

**First**, there appears to be a small typo in the agency table for the TxSubmission2 protocol:
![image](https://user-images.githubusercontent.com/49870/217457040-505959e6-be7a-4c88-8149-98105694b600.png)

I believe the client has agency during StTxs, not the server; this would match with the behavior I observed, and with the coloring on the diagram below it:
![image](https://user-images.githubusercontent.com/49870/217457166-a6d0836a-d70d-4971-bc8f-2d9e29c5fa7c.png)

**Second**, the CDDL for transaction IDs seems wrong.  It's left unspecified in the networking spec, which is fine, but the CDDL lists txID as an integer:

https://github.com/input-output-hk/ouroboros-network/blob/358880057f1d6d86cc385e8475813ace3b9ebef6/ouroboros-network-protocols/test-cddl/specs/common.cddl#L22

One might assume that this is a typo and it should obviously be a 32 byte hash of the transaction, I was surprised to learn this is not the case!

From a real mainnet node I'm getting data like this:
```
            [
                5,
                h'8c1c28940046ae8b8bab592936354ae833f99bc5b1768e07fefb0f7e617add2b',
            ],
```

My best guess is that the `5` is some kind of era signifier, but it'd be super helpful to make sure this is fully specified / documented somewhere.

**Comment by coot** [comment]
Thanks @Quantumplation.  Indeed `StTxs` should be in mentioned in the `Client has Agency` row.  As of `TxId`, in `ouroboros-network` we don't know what `Block`, `Tx` or `TxId` id is, that's why an `int` was choosen in the `cddl` spec (`cddl` does not have the notion of parametric polymorphism which we freely use in the Haskell code base).

But anyway, I think we should make this clearer in the technical report and we should at least point where this information can be found.

**Comment by Quantumplation** [comment]
@coot are there era-/protocol-specific versions of the CDDL for these polymorphic bits?

**Comment by coot** [comment]
I think ledger has it's own `CDDL` specs, but there's no place where we combine these into a single spec.  Ledger knows nothing about the network protocols, and we know very little about the ledger, while consensus the layer in the middle does not use `CDDL` at all (afaik).

**Comment by coot** [comment]
@JaredCorduan, @dnadales where should I point to for `ledger` `CDDL` for things like transaction `txid`, `tx` etc.  I actually don't know how far from truth are some of our arbitrary choices [here](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network-protocols/test-cddl/specs/common.cddl) - but I would like to add some useful comments / links there so people are not lost.

I think this kind of information should be written in the consensus report together with the HFC for `Cardano` - that would be a nice link to have, but I suspect it's one of the to be written sections since a long time :grin:.

**Comment by JaredCorduan** [comment]
> @JaredCorduan, @dnadales where should I point to for `ledger` `CDDL` for things like transaction `txid`, `tx` etc. I actually don't know how far from truth are some of our arbitrary choices [here](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network-protocols/test-cddl/specs/common.cddl) - but I would like to add some useful comments / links there so people are not lost.

That would be awesome!

The table at the top of the main ledger README has links to each era's CDDL files: https://github.com/input-output-hk/cardano-ledger#cardano-ledger

> I think this kind of information should be written in the consensus report together with the HFC for `Cardano` - that would be a nice link to have, but I suspect it's one of the to be written sections since a long time grin.

that would also be awesome!

**Comment by JaredCorduan** [comment]
In ledger, a `txid` is a 32-byte hash: https://github.com/input-output-hk/cardano-ledger/blob/ac405a977557a7c58ce1cf69d3c2a0bf148cf19f/eras/babbage/test-suite/cddl-files/babbage.cddl#L75

I'm guessing that the `5` wrapping comes from the HFC... perhaps it is referring to babbage as era number 5? (zero-indexing)

**Comment by Quantumplation** [comment]
Some observations as an outsider:
 - having network and ledger CDDL decoupled from each-other seems like a really good idea! That being said, it would be really cool to have some kind of build process that pulls-and-merges these CDDL specifications to a versioned, *single* consumable artifact for outside consumers; especially with the focus that other folks like dcSpark are putting on code-generation from CDDL, this would be very valuable.
 - in a similar vein, applying some unification to the naming across the two seems valuable; for example, `txid` in the network spec, vs `transaction_id` in the ledger spec (though I suspect these aren't the same, given the link that Jared provided!)

Would it be valuable to define some kind of `pCDDL` language, for "partial CDDL", that had support for "this type is explicitly left unspecified", with a toolchain to wire up related CDDL files?  Maybe such a thing already exists

**Comment by JaredCorduan** [comment]
> Would it be valuable to define some kind of `pCDDL` language, for "partial CDDL", that had support for "this type is explicitly left unspecified", with a toolchain to wire up related CDDL files? Maybe such a thing already exists

Here is something that we do in ledger, but it is admittedly more of a hack than a solution.

Our main CDDL file uses some undefined variables, such as `hash32` here: https://github.com/input-output-hk/cardano-ledger/blob/ac405a977557a7c58ce1cf69d3c2a0bf148cf19f/eras/conway/test-suite/cddl-files/conway.cddl#L459

Then we can have both `real` and `mock` definitions:
* https://github.com/input-output-hk/cardano-ledger/blob/ac405a977557a7c58ce1cf69d3c2a0bf148cf19f/eras/conway/test-suite/cddl-files/real/crypto.cddl#L2

Our property tests know how to cat the right files together to get a complete spec.

In general, I've found that CDDL does not have a lot of support.

**Comment by Quantumplation** [comment]
@JaredCorduan yea, that seems like kind of the flow I'm thinking, but then applied across multiple repositories to provide a per-era or per-hardfork unified specification as a consumable artifact for things like https://github.com/dcSpark/cardano-multiplatform-lib

As it stands, for the things it has defined, it seems like the CML has much more accurate specifications.

**Comment by JaredCorduan** [comment]
If you find anything different between the CML specs and the ledger CDDL, please let me know. We do property test the ledger CDDL specs, I'm fairly confident that they are not inaccurate (and I do talk with the CML folks whenever they find issues).  I don't think they do anything further up the stack than ledger.

**Comment by Quantumplation** [comment]
@JaredCorduan I'll tag in @SebastienGllmt for that :sweat_smile:

**Comment by JaredCorduan** [comment]
and don't forget @rooooooooob! :grin:

**Comment by rooooooooob** [comment]
Yeah CML is just the ledger-level CBOR-wise, plus utility functionality (cryptography, tx building/fees, etc) for creating those ledger objects. Since the CBOR part is 99% generated from the cddl spec there shouldn't be any differences between CML and the ledger cddl spec. A few times we've had small differences in the ledger's data vs the cddl spec but Jared has helped us quickly any time that has come up (e.g. small mistakes in cddl, or some details too specific to be expressed in cddl).

**Comment by Quantumplation** [comment]
@coot @JaredCorduan sorry to keep this thread around, but I found another underspecified bit that I can't figure out;  When I receive transactions back via the txsubmission miniprotocol, they come encoded like this:
```
        [
            5,
            24_0(<<[
                {tx body}
            ]>>),
        ],
```

The 5 here we decided is a 0-indexed era; what is the `24` tag signify? can it ever be anything else?  I couldn't find anything in either the network spec or the ledger spec to explain this.
